TypeScript defintion of StoreEnhancer does not allow state extension with replaceReducer

**Do you want to request a _feature_ or report a _bug_?**

The `StoreEnhancer` TypeScript interface does not seem to have the appropriate type for the `replaceReducer` property of the enhanced store. I would consider it a bug in the type definition.

**What is the current behavior?**

When implementing the `StoreEnhancer` interface with `ExtraState`, the type signature of `replaceReducer` on the "enhanced" store is coupled to the type of the wrapped reducer.

Given a `StoreEnhancer` of type

```ts
StoreEnhancer<Ext, ExtraState>
```

the returned store is of type

```ts
Store<S & StateExt, A> & Ext
```

with the `replaceReducer` property as such

```ts
(nextReducer: Reducer<S & ExtraState, A>) => void
```

Returning a store with a `replaceReducer` that accepts the original reducer gives the following type-error:

```
Type '<S, A extends Action<any> = AnyAction>(reducer: Reducer<S, A>, preloadedState?: DeepPartial<S> | undefined) => { replaceReducer: (nextReducer: Reducer<S, A>) => void; dispatch: Dispatch<A>; getState(): S & ExtraState; subscribe(listener: () => void): Unsubscribe; [Symbol.observable](): Observable<...>; }' is not assignable to type 'StoreEnhancerStoreCreator<{}, ExtraState>'.
      Type '{ replaceReducer: (nextReducer: Reducer<S, A>) => void; dispatch: Dispatch<A>; getState(): S & ExtraState; subscribe(listener: () => void): Unsubscribe; [Symbol.observable](): Observable<S & ExtraState>; }' is not assignable to type 'Store<S & ExtraState, A>'.
        Types of property 'replaceReducer' are incompatible.
          Type '(nextReducer: Reducer<S, A>) => void' is not assignable to type '(nextReducer: Reducer<S & ExtraState, A>) => void'.
            Types of parameters 'nextReducer' and 'nextReducer' are incompatible.
              Types of parameters 'state' and 'state' are incompatible.
                Type 'S | undefined' is not assignable to type '(S & ExtraState) | undefined'.
                  Type 'S' is not assignable to type '(S & ExtraState) | undefined'.
                    Type 'S' is not assignable to type 'S & ExtraState'.
                      Type 'S' is not assignable to type 'ExtraState'.

```

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://jsfiddle.net or similar.**

The type check fails for the following code:

```ts
import {
  StoreEnhancer,
  Action,
  AnyAction,
  Reducer,
  createStore,
  DeepPartial
} from "redux";

interface State {
  someField: "string";
}

interface ExtraState {
  extraField: "extra";
}

const reducer: Reducer<State> = null as any;

function stateExtensionExpectedToWork() {
  interface ExtraState {
    extraField: "extra";
  }

  const enhancer: StoreEnhancer<{}, ExtraState> = createStore => <
    S,
    A extends Action = AnyAction
  >(
    reducer: Reducer<S, A>,
    preloadedState?: DeepPartial<S>
  ) => {
    const wrappedReducer: Reducer<S & ExtraState, A> = null as any;
    const wrappedPreloadedState: S & ExtraState = null as any;
    const store = createStore(wrappedReducer, wrappedPreloadedState);
    return {
      ...store,
      replaceReducer: (nextReducer: Reducer<S, A>): void => {
        const nextWrappedReducer: Reducer<S & ExtraState, A> = null as any;
        store.replaceReducer(nextWrappedReducer);
      }
    };
  };

  const store = createStore(reducer, enhancer);
  store.replaceReducer(reducer);
}
```

See `src/index.ts` in the linked codesandbox example that implements the same function that would be expected to type-check, followed by another function of how it actually has to be done.

https://codesandbox.io/s/redux-store-enhancer-types-s6d3v

The example is based on the typescript test for the enhancer at [`test/typescript/enhancers.ts`](https://github.com/reduxjs/redux/blob/b6e1586d382b7e1f8f27776c3f422ef59efcd33a/test/typescript/enhancers.ts#L36) in this repo. The code doesn't execute (due to unsafe casts of null as any), but it is the type check that is of interest here.

**What is the expected behavior?**

When a store is created with a store enhancer that wraps a reducer and adds state, I would expect that `replaceReducer` on the returned store can be called with the original `rootReducer`.

```ts
const store = createStore(rootReducer, preloadedState, enhancer)
// ...
store.replaceReducer(rootReducer)
```

It would be the responsibility of the `enhancer` to appropriately replace the wrapped reducer. I.e return a store such as:

```ts
{
  ...store,
  replaceReducer: (nextReducer: Reducer<S, A>) => {
    store.replaceReducer(wrapReducer(nextReducer))
  },
}

```


**Which versions of Redux, and which browser and OS are affected by this issue? Did this work in previous versions of Redux?**

Redux version: 4.0.4,
OS: Ubuntu 19.04
Browser: N/A
Did this work in previous versions of Redux?: Not that I know of

@cellog: I tried against the branch, but I could not get it to work. I'm not sure if I'm missing something here..

```ts
import {
  StoreEnhancer,
  Action,
  AnyAction,
  Reducer,
  createStore,
  PreloadedState
} from "redux";

interface State {
  someField: "string";
}

interface ExtraState {
  extraField: "extra";
}

const reducer: Reducer<State> = null as any;

function stateExtensionExpectedToWork() {
  interface ExtraState {
    extraField: "extra";
  }

  const enhancer: StoreEnhancer<{}, ExtraState> = createStore => <
    S,
    A extends Action = AnyAction
  >(
    reducer: Reducer<S, A>,
    preloadedState?: PreloadedState<S>
  ) => {
    const wrappedReducer: Reducer<S & ExtraState, A> = null as any;
    const wrappedPreloadedState: PreloadedState<S & ExtraState> = null as any;
    const store = createStore(wrappedReducer, wrappedPreloadedState);
    return {
      ...store,
      replaceReducer: (nextReducer: Reducer<S, A>) => {
        const nextWrappedReducer: Reducer<S & ExtraState, A> = null as any;
        return store.replaceReducer(nextWrappedReducer);
      }
    };
  };

  const store = createStore(reducer, enhancer);
  store.replaceReducer(reducer);
}

```

I get the following error:

```
src/redux/store-enhancers/ext-enhancer.ts:25:66 - error TS2322: Type '<S, A extends Action<any> = AnyAction>(reducer: Reducer<S, A>, preloadedState?: PreloadedState<S>) => { replaceReducer: (nextReducer: Reducer<S, A>) => Store<S & ExtraState, A, ExtraState, {}>; dispatch: Dispatch<...>; getState(): S & ExtraState; subscribe(listener: () => void): Unsubscribe; [Symbol.observable](): O...' is not assignable to type 'StoreEnhancerStoreCreator<{}, ExtraState>'.
  Type '{ replaceReducer: (nextReducer: Reducer<S, A>) => Store<S & ExtraState, A, ExtraState, {}>; dispatch: Dispatch<A>; getState(): S & ExtraState; subscribe(listener: () => void): Unsubscribe; [Symbol.observable](): Observable<...>; }' is not assignable to type 'Store<S & ExtraState, A, ExtraState, {}>'.
    Types of property 'replaceReducer' are incompatible.
      Type '(nextReducer: Reducer<S, A>) => Store<S & ExtraState, A, ExtraState, {}>' is not assignable to type '<NewState = S & ExtraState, NewActions extends A = A>(nextReducer: Reducer<NewState, NewActions>) => Store<NewState & ExtraState, NewActions, ExtraState, {}>'.
        Types of parameters 'nextReducer' and 'nextReducer' are incompatible.
          Types of parameters 'state' and 'state' are incompatible.
            Type 'S' is not assignable to type 'NewState'.
              'S' is assignable to the constraint of type 'NewState', but 'NewState' could be instantiated with a different subtype of constraint '{}'.

 25   const enhancer: StoreEnhancer<{}, ExtraState> = createStore => <
                                                                     ~
 26     S,
    ~~~~~~
...
 30     preloadedState?: PreloadedState<S>
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 31   ) => {
    ~~~~~~~~

  ../redux/index.d.ts:436:54
    436 export type StoreEnhancer<Ext = {}, StateExt = {}> = (
                                                             ~
    437   next: StoreEnhancerStoreCreator<Ext, StateExt>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    438 ) => StoreEnhancerStoreCreator<Ext, StateExt>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The expected type comes from the return type of this signature.


Found 1 error.
```

--- a/index.d.ts
+++ b/index.d.ts
@@ -260,7 +260,7 @@ export type Observer<T> = {
 export interface Store<
   S = any,
   A extends Action = AnyAction,
-  StateExt = S,
+  StateExt = S extends {} ? {} : S extends [] ? [] : S,
   Ext = {}
 > {
   /**
@@ -362,11 +362,21 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext = {}, StateExt = S>(
+  <
+    S,
+    A extends Action,
+    Ext = {},
+    StateExt = S extends {} ? {} : S extends [] ? [] : S
+  >(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
   ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext = {}, StateExt = S>(
+  <
+    S,
+    A extends Action,
+    Ext = {},
+    StateExt = S extends {} ? {} : S extends [] ? [] : S
+  >(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>

--- a/index.d.ts
+++ b/index.d.ts
@@ -362,11 +362,11 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext extends {}, StateExt>(
+  <S, A extends Action, Ext, StateExt>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
   ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext extends {}, StateExt>(
+  <S, A extends Action, Ext, StateExt>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>

--- a/index.d.ts
+++ b/index.d.ts
@@ -260,7 +260,7 @@ export type Observer<T> = {
 export interface Store<
   S = any,
   A extends Action = AnyAction,
-  StateExt = {},
+  StateExt = S,
   Ext = {}
 > {
   /**
@@ -362,11 +362,11 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext = {}, StateExt = {}>(
+  <S, A extends Action, Ext = {}, StateExt = S>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
   ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext = {}, StateExt = {}>(
+  <S, A extends Action, Ext = {}, StateExt = S>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>

--- a/index.d.ts
+++ b/index.d.ts
@@ -362,11 +362,11 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext, StateExt>(
+  <S, A extends Action, Ext = {}, StateExt = {}>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
   ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext, StateExt>(
+  <S, A extends Action, Ext = {}, StateExt = {}>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>

--- a/index.d.ts
+++ b/index.d.ts
@@ -248,13 +248,15 @@ export type Observer<T> = {
 }
 
 /**
- * returns the most basic type that will not interfere with the existing type
+ * Extend the state
  *
- * Note that for non-object stuff, this will mess with replaceReducer. The
- * assumption is that root reducers that only return a basic type (string, number, null, symbol) probably won't
- * be using replaceReducer anyways.
+ * This is used by store enhancers and store creators to extend state.
+ * If there is no state extension, it just returns the state, as is, otherwise
+ * it returns the state joined with its extension.
  */
-export type BaseType<S> = S extends {} ? {} : S
+export type ExtendState<State, Extension> = [Extension] extends [never]
+  ? State
+  : State & Extension
 
 /**
  * A store is an object that holds the application's state tree.
@@ -269,7 +271,7 @@ export type BaseType<S> = S extends {} ? {} : S
 export interface Store<
   S = any,
   A extends Action = AnyAction,
-  StateExt = BaseType<S>,
+  StateExt = never,
   Ext = {}
 > {
   /**
@@ -344,7 +346,7 @@ export interface Store<
    */
   replaceReducer<NewState, NewActions extends Action>(
     nextReducer: Reducer<NewState, NewActions>
-  ): Store<NewState & StateExt, NewActions, StateExt, Ext> & Ext
+  ): Store<ExtendState<NewState, StateExt>, NewActions, StateExt, Ext> & Ext
 
   /**
    * Interoperability point for observable/reactive libraries.
@@ -371,15 +373,15 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
+  <S, A extends Action, Ext = {}, StateExt = never>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
-  ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
+  ): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
+  <S, A extends Action, Ext = {}, StateExt = never>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>
-  ): Store<S & StateExt, A, StateExt, Ext> & Ext
+  ): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
 }
 
 /**
@@ -433,17 +435,17 @@ export const createStore: StoreCreator
  * @template Ext Store extension that is mixed into the Store type.
  * @template StateExt State extension that is mixed into the state type.
  */
-export type StoreEnhancer<Ext = {}, StateExt = {}> = (
+export type StoreEnhancer<Ext = {}, StateExt = never> = (
   next: StoreEnhancerStoreCreator<Ext, StateExt>
 ) => StoreEnhancerStoreCreator<Ext, StateExt>
 
-export type StoreEnhancerStoreCreator<Ext = {}, StateExt = {}> = <
+export type StoreEnhancerStoreCreator<Ext = {}, StateExt = never> = <
   S = any,
   A extends Action = AnyAction
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>
-) => Store<S & StateExt, A, StateExt, Ext> & Ext
+) => Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
 
 /* middleware */
 

--- a/index.d.ts
+++ b/index.d.ts
@@ -362,11 +362,11 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext extends {}, StateExt extends {}>(
+  <S, A extends Action, Ext extends {}, StateExt>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
   ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext extends {}, StateExt extends {}>(
+  <S, A extends Action, Ext extends {}, StateExt>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>

--- a/index.d.ts
+++ b/index.d.ts
@@ -254,8 +254,15 @@ export type Observer<T> = {
  *
  * @template S The type of state held by this store.
  * @template A the type of actions which may be dispatched by this store.
+ * @template StateExt any extension to state from store enhancers
+ * @template Ext any extensions to the store from store enhancers
  */
-export interface Store<S = any, A extends Action = AnyAction> {
+export interface Store<
+  S = any,
+  A extends Action = AnyAction,
+  StateExt = {},
+  Ext = {}
+> {
   /**
    * Dispatches an action. It is the only way to trigger a state change.
    *
@@ -328,7 +335,7 @@ export interface Store<S = any, A extends Action = AnyAction> {
    */
   replaceReducer<NewState = S, NewActions extends A = A>(
     nextReducer: Reducer<NewState, NewActions>
-  ): Store<NewState, NewActions>
+  ): Store<NewState & StateExt, NewActions, StateExt, Ext> & Ext
 
   /**
    * Interoperability point for observable/reactive libraries.
@@ -355,15 +362,15 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext, StateExt>(
+  <S, A extends Action, Ext extends {}, StateExt extends {}>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
-  ): Store<S & StateExt, A> & Ext
-  <S, A extends Action, Ext, StateExt>(
+  ): Store<S & StateExt, A, StateExt, Ext> & Ext
+  <S, A extends Action, Ext extends {}, StateExt extends {}>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>
-  ): Store<S & StateExt, A> & Ext
+  ): Store<S & StateExt, A, StateExt, Ext> & Ext
 }
 
 /**
@@ -418,15 +425,16 @@ export const createStore: StoreCreator
  * @template StateExt State extension that is mixed into the state type.
  */
 export type StoreEnhancer<Ext = {}, StateExt = {}> = (
-  next: StoreEnhancerStoreCreator
+  next: StoreEnhancerStoreCreator<Ext, StateExt>
 ) => StoreEnhancerStoreCreator<Ext, StateExt>
+
 export type StoreEnhancerStoreCreator<Ext = {}, StateExt = {}> = <
   S = any,
   A extends Action = AnyAction
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>
-) => Store<S & StateExt, A> & Ext
+) => Store<S & StateExt, A, StateExt, Ext> & Ext
 
 /* middleware */
 

--- a/index.d.ts
+++ b/index.d.ts
@@ -342,7 +342,7 @@ export interface Store<
    *
    * @param nextReducer The reducer for the store to use instead.
    */
-  replaceReducer<NewState = S, NewActions extends A = A>(
+  replaceReducer<NewState, NewActions extends Action>(
     nextReducer: Reducer<NewState, NewActions>
   ): Store<NewState & StateExt, NewActions, StateExt, Ext> & Ext
 

--- a/index.d.ts
+++ b/index.d.ts
@@ -248,6 +248,15 @@ export type Observer<T> = {
 }
 
 /**
+ * returns the most basic type that will not interfere with the existing type
+ *
+ * Note that for non-object stuff, this will mess with replaceReducer. The
+ * assumption is that root reducers that only return a basic type (string, number, null, symbol) probably won't
+ * be using replaceReducer anyways.
+ */
+export type BaseType<S> = S extends {} ? {} : S
+
+/**
  * A store is an object that holds the application's state tree.
  * There should only be a single store in a Redux app, as the composition
  * happens on the reducer level.
@@ -260,7 +269,7 @@ export type Observer<T> = {
 export interface Store<
   S = any,
   A extends Action = AnyAction,
-  StateExt = S extends {} ? {} : S extends [] ? [] : S,
+  StateExt = BaseType<S>,
   Ext = {}
 > {
   /**
@@ -362,21 +371,11 @@ export type DeepPartial<T> = {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <
-    S,
-    A extends Action,
-    Ext = {},
-    StateExt = S extends {} ? {} : S extends [] ? [] : S
-  >(
+  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
   ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <
-    S,
-    A extends Action,
-    Ext = {},
-    StateExt = S extends {} ? {} : S extends [] ? [] : S
-  >(
+  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>

--- a/src/applyMiddleware.ts
+++ b/src/applyMiddleware.ts
@@ -51,7 +51,7 @@ export default function applyMiddleware<Ext, S = any>(
 ): StoreEnhancer<{ dispatch: Ext }>
 export default function applyMiddleware(
   ...middlewares: Middleware[]
-): StoreEnhancer {
+): StoreEnhancer<any> {
   return (createStore: StoreCreator) => <S, A extends AnyAction>(
     reducer: Reducer<S, A>,
     ...args: any[]

--- a/src/createStore.ts
+++ b/src/createStore.ts
@@ -6,7 +6,7 @@ import {
   StoreEnhancer,
   Dispatch,
   Observer,
-  BaseType
+  ExtendState
 } from './types/store'
 import { Action } from './types/actions'
 import { Reducer } from './types/reducers'
@@ -42,31 +42,31 @@ export default function createStore<
   S,
   A extends Action,
   Ext = {},
-  StateExt = BaseType<S>
+  StateExt = never
 >(
   reducer: Reducer<S, A>,
   enhancer?: StoreEnhancer<Ext, StateExt>
-): Store<S & StateExt, A, StateExt, Ext> & Ext
+): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
 export default function createStore<
   S,
   A extends Action,
   Ext = {},
-  StateExt = BaseType<S>
+  StateExt = never
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>,
   enhancer?: StoreEnhancer<Ext>
-): Store<S & StateExt, A, StateExt, Ext> & Ext
+): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
 export default function createStore<
   S,
   A extends Action,
   Ext = {},
-  StateExt = BaseType<S>
+  StateExt = never
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S> | StoreEnhancer<Ext, StateExt>,
   enhancer?: StoreEnhancer<Ext, StateExt>
-): Store<S & StateExt, A, StateExt, Ext> & Ext {
+): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext {
   if (
     (typeof preloadedState === 'function' && typeof enhancer === 'function') ||
     (typeof enhancer === 'function' && typeof arguments[3] === 'function')
@@ -90,7 +90,7 @@ export default function createStore<
 
     return enhancer(createStore)(reducer, preloadedState as PreloadedState<
       S
-    >) as Store<S & StateExt, A, StateExt, Ext> & Ext
+    >) as Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
   }
 
   if (typeof reducer !== 'function') {
@@ -268,7 +268,7 @@ export default function createStore<
    */
   function replaceReducer<NewState, NewActions extends A>(
     nextReducer: Reducer<NewState, NewActions>
-  ): Store<NewState, NewActions> {
+  ): Store<ExtendState<NewState, StateExt>, NewActions, StateExt, Ext> & Ext {
     if (typeof nextReducer !== 'function') {
       throw new Error('Expected the nextReducer to be a function.')
     }
@@ -285,7 +285,13 @@ export default function createStore<
     // the new state tree with any relevant data from the old one.
     dispatch({ type: ActionTypes.REPLACE } as A)
     // change the type of the store by casting it to the new store
-    return (store as unknown) as Store<NewState, NewActions>
+    return (store as unknown) as Store<
+      ExtendState<NewState, StateExt>,
+      NewActions,
+      StateExt,
+      Ext
+    > &
+      Ext
   }
 
   /**
@@ -339,6 +345,6 @@ export default function createStore<
     getState,
     replaceReducer,
     [$$observable]: observable
-  } as unknown) as Store<S & StateExt, A, StateExt, Ext> & Ext
+  } as unknown) as Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
   return store
 }

--- a/src/createStore.ts
+++ b/src/createStore.ts
@@ -333,12 +333,12 @@ export default function createStore<
   // the initial state tree.
   dispatch({ type: ActionTypes.INIT } as A)
 
-  const store: Store<S & StateExt, A> & Ext = ({
+  const store = ({
     dispatch: dispatch as Dispatch<A>,
     subscribe,
     getState,
     replaceReducer,
     [$$observable]: observable
-  } as unknown) as Store<S & StateExt, A> & Ext
+  } as unknown) as Store<S & StateExt, A, StateExt, Ext> & Ext
   return store
 }

--- a/src/createStore.ts
+++ b/src/createStore.ts
@@ -55,7 +55,7 @@ export default function createStore<
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>,
-  enhancer?: StoreEnhancer<Ext>
+  enhancer?: StoreEnhancer<Ext, StateExt>
 ): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
 export default function createStore<
   S,

--- a/src/createStore.ts
+++ b/src/createStore.ts
@@ -90,7 +90,7 @@ export default function createStore<
 
     return enhancer(createStore)(reducer, preloadedState as PreloadedState<
       S
-    >) as Store<S & StateExt, A> & Ext
+    >) as Store<S & StateExt, A, StateExt, Ext> & Ext
   }
 
   if (typeof reducer !== 'function') {

--- a/src/createStore.ts
+++ b/src/createStore.ts
@@ -5,7 +5,8 @@ import {
   PreloadedState,
   StoreEnhancer,
   Dispatch,
-  Observer
+  Observer,
+  BaseType
 } from './types/store'
 import { Action } from './types/actions'
 import { Reducer } from './types/reducers'
@@ -37,20 +38,35 @@ import isPlainObject from './utils/isPlainObject'
  * @returns {Store} A Redux store that lets you read the state, dispatch actions
  * and subscribe to changes.
  */
-export default function createStore<S, A extends Action, Ext, StateExt>(
+export default function createStore<
+  S,
+  A extends Action,
+  Ext = {},
+  StateExt = BaseType<S>
+>(
   reducer: Reducer<S, A>,
   enhancer?: StoreEnhancer<Ext, StateExt>
-): Store<S & StateExt, A> & Ext
-export default function createStore<S, A extends Action, Ext, StateExt>(
+): Store<S & StateExt, A, StateExt, Ext> & Ext
+export default function createStore<
+  S,
+  A extends Action,
+  Ext = {},
+  StateExt = BaseType<S>
+>(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>,
   enhancer?: StoreEnhancer<Ext>
-): Store<S & StateExt, A> & Ext
-export default function createStore<S, A extends Action, Ext, StateExt>(
+): Store<S & StateExt, A, StateExt, Ext> & Ext
+export default function createStore<
+  S,
+  A extends Action,
+  Ext = {},
+  StateExt = BaseType<S>
+>(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S> | StoreEnhancer<Ext, StateExt>,
-  enhancer?: StoreEnhancer<Ext>
-): Store<S & StateExt, A> & Ext {
+  enhancer?: StoreEnhancer<Ext, StateExt>
+): Store<S & StateExt, A, StateExt, Ext> & Ext {
   if (
     (typeof preloadedState === 'function' && typeof enhancer === 'function') ||
     (typeof enhancer === 'function' && typeof arguments[3] === 'function')

--- a/src/index.ts
+++ b/src/index.ts
@@ -19,7 +19,8 @@ export {
   Store,
   StoreCreator,
   StoreEnhancer,
-  StoreEnhancerStoreCreator
+  StoreEnhancerStoreCreator,
+  ExtendState
 } from './types/store'
 // reducers
 export {

--- a/src/types/store.ts
+++ b/src/types/store.ts
@@ -8,6 +8,9 @@ import { Reducer } from './reducers'
  * This is used by store enhancers and store creators to extend state.
  * If there is no state extension, it just returns the state, as is, otherwise
  * it returns the state joined with its extension.
+ *
+ * Reference for future devs:
+ * https://github.com/microsoft/TypeScript/issues/31751#issuecomment-498526919
  */
 export type ExtendState<State, Extension> = [Extension] extends [never]
   ? State

--- a/src/types/store.ts
+++ b/src/types/store.ts
@@ -3,13 +3,15 @@ import { Action, AnyAction } from './actions'
 import { Reducer } from './reducers'
 
 /**
- * returns the most basic type that will not interfere with the existing type
+ * Extend the state
  *
- * Note that for non-object stuff, this will mess with replaceReducer. The
- * assumption is that root reducers that only return a basic type (string, number, null, symbol) probably won't
- * be using replaceReducer anyways.
+ * This is used by store enhancers and store creators to extend state.
+ * If there is no state extension, it just returns the state, as is, otherwise
+ * it returns the state joined with its extension.
  */
-export type BaseType<S> = S extends {} ? {} : S
+export type ExtendState<State, Extension> = [Extension] extends [never]
+  ? State
+  : State & Extension
 
 /**
  * Internal "virtual" symbol used to make the `CombinedState` type unique.
@@ -122,7 +124,7 @@ export type Observer<T> = {
 export interface Store<
   S = any,
   A extends Action = AnyAction,
-  StateExt = BaseType<S>,
+  StateExt = never,
   Ext = {}
 > {
   /**
@@ -197,7 +199,7 @@ export interface Store<
    */
   replaceReducer<NewState, NewActions extends Action>(
     nextReducer: Reducer<NewState, NewActions>
-  ): Store<NewState & StateExt, NewActions, StateExt, Ext> & Ext
+  ): Store<ExtendState<NewState, StateExt>, NewActions, StateExt, Ext> & Ext
 
   /**
    * Interoperability point for observable/reactive libraries.
@@ -220,15 +222,15 @@ export interface Store<
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
+  <S, A extends Action, Ext = {}, StateExt = never>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
-  ): Store<S & StateExt, A, StateExt, Ext> & Ext
-  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
+  ): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
+  <S, A extends Action, Ext = {}, StateExt = never>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>
-  ): Store<S & StateExt, A, StateExt, Ext> & Ext
+  ): Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext
 }
 
 /**
@@ -261,4 +263,4 @@ export type StoreEnhancerStoreCreator<Ext = {}, StateExt = {}> = <
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>
-) => Store<S & StateExt, A, StateExt, Ext> & Ext
+) => Store<ExtendState<S, StateExt>, A, StateExt, Ext> & Ext

--- a/src/types/store.ts
+++ b/src/types/store.ts
@@ -3,6 +3,15 @@ import { Action, AnyAction } from './actions'
 import { Reducer } from './reducers'
 
 /**
+ * returns the most basic type that will not interfere with the existing type
+ *
+ * Note that for non-object stuff, this will mess with replaceReducer. The
+ * assumption is that root reducers that only return a basic type (string, number, null, symbol) probably won't
+ * be using replaceReducer anyways.
+ */
+export type BaseType<S> = S extends {} ? {} : S
+
+/**
  * Internal "virtual" symbol used to make the `CombinedState` type unique.
  */
 declare const $CombinedState: unique symbol
@@ -107,8 +116,15 @@ export type Observer<T> = {
  *
  * @template S The type of state held by this store.
  * @template A the type of actions which may be dispatched by this store.
+ * @template StateExt any extension to state from store enhancers
+ * @template Ext any extensions to the store from store enhancers
  */
-export interface Store<S = any, A extends Action = AnyAction> {
+export interface Store<
+  S = any,
+  A extends Action = AnyAction,
+  StateExt = BaseType<S>,
+  Ext = {}
+> {
   /**
    * Dispatches an action. It is the only way to trigger a state change.
    *
@@ -179,9 +195,9 @@ export interface Store<S = any, A extends Action = AnyAction> {
    *
    * @param nextReducer The reducer for the store to use instead.
    */
-  replaceReducer<NewState = S, NewActions extends A = A>(
+  replaceReducer<NewState, NewActions extends Action>(
     nextReducer: Reducer<NewState, NewActions>
-  ): Store<NewState, NewActions>
+  ): Store<NewState & StateExt, NewActions, StateExt, Ext> & Ext
 
   /**
    * Interoperability point for observable/reactive libraries.
@@ -204,15 +220,15 @@ export interface Store<S = any, A extends Action = AnyAction> {
  * @template StateExt State extension that is mixed into the state type.
  */
 export interface StoreCreator {
-  <S, A extends Action, Ext, StateExt>(
+  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
     reducer: Reducer<S, A>,
     enhancer?: StoreEnhancer<Ext, StateExt>
-  ): Store<S & StateExt, A> & Ext
-  <S, A extends Action, Ext, StateExt>(
+  ): Store<S & StateExt, A, StateExt, Ext> & Ext
+  <S, A extends Action, Ext = {}, StateExt = BaseType<S>>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>,
     enhancer?: StoreEnhancer<Ext>
-  ): Store<S & StateExt, A> & Ext
+  ): Store<S & StateExt, A, StateExt, Ext> & Ext
 }
 
 /**
@@ -237,7 +253,7 @@ export interface StoreCreator {
  * @template StateExt State extension that is mixed into the state type.
  */
 export type StoreEnhancer<Ext = {}, StateExt = {}> = (
-  next: StoreEnhancerStoreCreator
+  next: StoreEnhancerStoreCreator<Ext, StateExt>
 ) => StoreEnhancerStoreCreator<Ext, StateExt>
 export type StoreEnhancerStoreCreator<Ext = {}, StateExt = {}> = <
   S = any,
@@ -245,4 +261,4 @@ export type StoreEnhancerStoreCreator<Ext = {}, StateExt = {}> = <
 >(
   reducer: Reducer<S, A>,
   preloadedState?: PreloadedState<S>
-) => Store<S & StateExt, A> & Ext
+) => Store<S & StateExt, A, StateExt, Ext> & Ext

--- a/src/types/store.ts
+++ b/src/types/store.ts
@@ -254,10 +254,10 @@ export interface StoreCreator {
  * @template Ext Store extension that is mixed into the Store type.
  * @template StateExt State extension that is mixed into the state type.
  */
-export type StoreEnhancer<Ext = {}, StateExt = {}> = (
+export type StoreEnhancer<Ext = {}, StateExt = never> = (
   next: StoreEnhancerStoreCreator<Ext, StateExt>
 ) => StoreEnhancerStoreCreator<Ext, StateExt>
-export type StoreEnhancerStoreCreator<Ext = {}, StateExt = {}> = <
+export type StoreEnhancerStoreCreator<Ext = {}, StateExt = never> = <
   S = any,
   A extends Action = AnyAction
 >(

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -80,10 +80,10 @@ function replaceReducerExtender() {
     extraField: 'extra'
   }
 
-  const enhancer: StoreEnhancer<{}, ExtraState> = createStore => <
-    S,
-    A extends Action = AnyAction
-  >(
+  const enhancer: StoreEnhancer<
+    { method(): string },
+    ExtraState
+  > = createStore => <S, A extends Action = AnyAction>(
     reducer: Reducer<S, A>,
     preloadedState?: PreloadedState<S>
   ) => {
@@ -104,4 +104,8 @@ function replaceReducerExtender() {
   store.getState().extraField
   // typings:expect-error
   store.getState().wrongField
+
+  const res: string = store.method()
+  // typings:expect-error
+  store.wrongMethod()
 }

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -131,18 +131,43 @@ function mhelmersonExample() {
       A extends Action = AnyAction
     >(
       reducer: Reducer<S, A>,
-      preloadedState?: PreloadedState<S>
+      preloadedState?: any
     ) => {
-      const wrappedReducer: Reducer<S & ExtraState, A> = null as any
-      const wrappedPreloadedState: PreloadedState<S & ExtraState> = null as any
+      const wrappedReducer = (state: S & ExtraState | undefined, action: A) => {
+        const newState = reducer(state, action)
+        return {
+          ...newState,
+          extraField: 'extra'
+        } as S & ExtraState
+      }
+      const wrappedPreloadedState = preloadedState
+        ? {
+            ...preloadedState,
+            extraField: 'extra'
+          }
+        : undefined
       const store = createStore(wrappedReducer, wrappedPreloadedState)
       return {
         ...store,
-        replaceReducer: (nextReducer: Reducer<S, A>) => {
-          const nextWrappedReducer: Reducer<S & ExtraState, A> = null as any
+        replaceReducer<NS, NA extends Action = AnyAction>(
+          nextReducer: (
+            state: NS & ExtraState | undefined,
+            action: NA
+          ) => NS & ExtraState
+        ) {
+          const nextWrappedReducer: Reducer<NS & ExtraState, NA> = (
+            state,
+            action
+          ) => {
+            const newState = nextReducer(state, action)
+            return {
+              ...newState,
+              extraField: 'extra'
+            }
+          }
           return store.replaceReducer(nextWrappedReducer)
         }
-      } as Store<S & ExtraState, A, ExtraState>
+      }
     }
 
     const store = createStore(reducer, enhancer)

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -71,3 +71,37 @@ function extraMethods() {
   // typings:expect-error
   store.wrongMethod()
 }
+
+/**
+ * replaceReducer with a store enhancer
+ */
+function replaceReducerExtender() {
+  interface ExtraState {
+    extraField: 'extra'
+  }
+
+  const enhancer: StoreEnhancer<{}, ExtraState> = createStore => <
+    S,
+    A extends Action = AnyAction
+  >(
+    reducer: Reducer<S, A>,
+    preloadedState?: PreloadedState<S>
+  ) => {
+    const wrappedReducer: Reducer<S & ExtraState, A> = null as any
+    const wrappedPreloadedState: PreloadedState<S & ExtraState> = null as any
+    return createStore(wrappedReducer, wrappedPreloadedState)
+  }
+
+  const store = createStore(reducer, enhancer)
+
+  const newReducer = (
+    state: { test: boolean } = { test: true },
+    _: AnyAction
+  ) => state
+
+  const newStore = store.replaceReducer(newReducer)
+  newStore.getState().test
+  store.getState().extraField
+  // typings:expect-error
+  store.getState().wrongField
+}

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -141,13 +141,13 @@ function replaceReducerExtender() {
 
   const newStore = store.replaceReducer(newReducer)
   newStore.getState().test
-  store.getState().extraField
+  newStore.getState().extraField
   // typings:expect-error
-  store.getState().wrongField
+  newStore.getState().wrongField
 
-  const res: string = store.method()
+  const res: string = newStore.method()
   // typings:expect-error
-  store.wrongMethod()
+  newStore.wrongMethod()
 }
 
 function mhelmersonExample() {
@@ -212,24 +212,56 @@ function mhelmersonExample() {
 
     const store = createStore(reducer, enhancer)
     store.replaceReducer(reducer)
+
+    store.getState().extraField
+    // typings:expect-error
+    store.getState().wrongField
+    // typings:expect-error
+    store.getState().test
+
+    const newReducer = (
+      state: { test: boolean } = { test: true },
+      _: AnyAction
+    ) => state
+
+    const newStore = store.replaceReducer(newReducer)
+    newStore.getState().test
+    newStore.getState().extraField
+    // typings:expect-error
+    newStore.getState().wrongField
   }
 }
 
 function finalHelmersonExample() {
-  function persistReducer<S>(config: any, reducer: S): S {
-    return reducer
+  interface ExtraState {
+    foo: string
+  }
+
+  function persistReducer<S, A extends AnyAction>(
+    config: any,
+    reducer: Reducer<S, A>
+  ) {
+    return (state: S & ExtraState | undefined, action: AnyAction) => {
+      const newState = reducer(state, (action as unknown) as A)
+      return {
+        ...newState,
+        foo: 'hi'
+      }
+    }
   }
 
   function persistStore<S>(store: S) {
     return store
   }
 
-  function createPersistEnhancer(persistConfig: any): StoreEnhancer {
+  function createPersistEnhancer(
+    persistConfig: any
+  ): StoreEnhancer<{}, ExtraState> {
     return createStore => <S, A extends Action = AnyAction>(
       reducer: Reducer<S, A>,
       preloadedState?: any
     ) => {
-      const persistedReducer = persistReducer(persistConfig, reducer)
+      const persistedReducer = persistReducer<S, A>(persistConfig, reducer)
       const store = createStore(persistedReducer, preloadedState)
       const persistor = persistStore(store)
 
@@ -244,4 +276,22 @@ function finalHelmersonExample() {
       }
     }
   }
+
+  const store = createStore(reducer, createPersistEnhancer('hi'))
+
+  store.getState().foo
+  // typings:expect-error
+  store.getState().wrongField
+
+  const newReducer = (
+    state: { test: boolean } = { test: true },
+    _: AnyAction
+  ) => state
+
+  const newStore = store.replaceReducer(newReducer)
+  newStore.getState().test
+  // typings:expect-error
+  newStore.getState().whatever
+  // typings:expect-error
+  newStore.getState().wrongField
 }

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -214,3 +214,34 @@ function mhelmersonExample() {
     store.replaceReducer(reducer)
   }
 }
+
+function finalHelmersonExample() {
+  function persistReducer<S>(config: any, reducer: S): S {
+    return reducer
+  }
+
+  function persistStore<S>(store: S) {
+    return store
+  }
+
+  function createPersistEnhancer(persistConfig: any): StoreEnhancer {
+    return createStore => <S, A extends Action = AnyAction>(
+      reducer: Reducer<S, A>,
+      preloadedState?: any
+    ) => {
+      const persistedReducer = persistReducer(persistConfig, reducer)
+      const store = createStore(persistedReducer, preloadedState)
+      const persistor = persistStore(store)
+
+      return {
+        ...store,
+        replaceReducer: nextReducer => {
+          return store.replaceReducer(
+            persistReducer(persistConfig, nextReducer)
+          )
+        },
+        persistor
+      }
+    }
+  }
+}

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -109,3 +109,43 @@ function replaceReducerExtender() {
   // typings:expect-error
   store.wrongMethod()
 }
+
+function mhelmersonExample() {
+  interface State {
+    someField: 'string'
+  }
+
+  interface ExtraState {
+    extraField: 'extra'
+  }
+
+  const reducer: Reducer<State> = null as any
+
+  function stateExtensionExpectedToWork() {
+    interface ExtraState {
+      extraField: 'extra'
+    }
+
+    const enhancer: StoreEnhancer<{}, ExtraState> = createStore => <
+      S,
+      A extends Action = AnyAction
+    >(
+      reducer: Reducer<S, A>,
+      preloadedState?: PreloadedState<S>
+    ) => {
+      const wrappedReducer: Reducer<S & ExtraState, A> = null as any
+      const wrappedPreloadedState: PreloadedState<S & ExtraState> = null as any
+      const store = createStore(wrappedReducer, wrappedPreloadedState)
+      return {
+        ...store,
+        replaceReducer: (nextReducer: Reducer<S, A>) => {
+          const nextWrappedReducer: Reducer<S & ExtraState, A> = null as any
+          return store.replaceReducer(nextWrappedReducer)
+        }
+      } as Store<S & ExtraState, A, ExtraState>
+    }
+
+    const store = createStore(reducer, enhancer)
+    store.replaceReducer(reducer)
+  }
+}

--- a/test/typescript/enhancers.ts
+++ b/test/typescript/enhancers.ts
@@ -18,7 +18,25 @@ const reducer: Reducer<State> = null as any
 function dispatchExtension() {
   type PromiseDispatch = <T extends Action>(promise: Promise<T>) => Promise<T>
 
-  const enhancer: StoreEnhancer<{ dispatch: PromiseDispatch }> = null as any
+  const enhancer: StoreEnhancer<{
+    dispatch: PromiseDispatch
+  }> = createStore => <S, A extends Action = AnyAction>(
+    reducer: Reducer<S, A>,
+    preloadedState?: any
+  ) => {
+    const store = createStore(reducer, preloadedState)
+    return {
+      ...store,
+      dispatch: (action: any) => {
+        if (action.type) {
+          store.dispatch(action)
+        } else if (action.then) {
+          action.then(store.dispatch)
+        }
+        return action
+      }
+    }
+  }
 
   const store = createStore(reducer, enhancer)
 
@@ -43,10 +61,21 @@ function stateExtension() {
     A extends Action = AnyAction
   >(
     reducer: Reducer<S, A>,
-    preloadedState?: PreloadedState<S>
+    preloadedState?: any
   ) => {
-    const wrappedReducer: Reducer<S & ExtraState, A> = null as any
-    const wrappedPreloadedState: PreloadedState<S & ExtraState> = null as any
+    const wrappedReducer: Reducer<S & ExtraState, A> = (state, action) => {
+      const newState = reducer(state, action)
+      return {
+        ...newState,
+        extraField: 'extra'
+      }
+    }
+    const wrappedPreloadedState = preloadedState
+      ? {
+          ...preloadedState,
+          extraField: 'extra'
+        }
+      : undefined
     return createStore(wrappedReducer, wrappedPreloadedState)
   }
 
@@ -85,10 +114,21 @@ function replaceReducerExtender() {
     ExtraState
   > = createStore => <S, A extends Action = AnyAction>(
     reducer: Reducer<S, A>,
-    preloadedState?: PreloadedState<S>
+    preloadedState?: any
   ) => {
-    const wrappedReducer: Reducer<S & ExtraState, A> = null as any
-    const wrappedPreloadedState: PreloadedState<S & ExtraState> = null as any
+    const wrappedReducer: Reducer<S & ExtraState, A> = (state, action) => {
+      const newState = reducer(state, action)
+      return {
+        ...newState,
+        extraField: 'extra'
+      }
+    }
+    const wrappedPreloadedState = preloadedState
+      ? {
+          ...preloadedState,
+          extraField: 'extra'
+        }
+      : undefined
     return createStore(wrappedReducer, wrappedPreloadedState)
   }
 
@@ -133,12 +173,12 @@ function mhelmersonExample() {
       reducer: Reducer<S, A>,
       preloadedState?: any
     ) => {
-      const wrappedReducer = (state: S & ExtraState | undefined, action: A) => {
+      const wrappedReducer: Reducer<S & ExtraState, A> = (state, action) => {
         const newState = reducer(state, action)
         return {
           ...newState,
           extraField: 'extra'
-        } as S & ExtraState
+        }
       }
       const wrappedPreloadedState = preloadedState
         ? {

--- a/test/typescript/store.ts
+++ b/test/typescript/store.ts
@@ -56,6 +56,9 @@ const funcWithStore = (store: Store<State, DerivedAction>) => {}
 
 const store: Store<State> = createStore(reducer)
 
+// ensure that an array-based state works
+const arrayReducer = (state = []) => state
+const storeWithArrayState: Store<[]> = createStore(arrayReducer)
 const storeWithPreloadedState: Store<State> = createStore(reducer, {
   a: 'a',
   b: { c: 'c', d: 'd' }

--- a/test/typescript/store.ts
+++ b/test/typescript/store.ts
@@ -7,7 +7,8 @@ import {
   StoreCreator,
   StoreEnhancerStoreCreator,
   Unsubscribe,
-  Observer
+  Observer,
+  ExtendState
 } from 'redux'
 import 'symbol-observable'
 
@@ -15,6 +16,41 @@ type State = {
   a: 'a'
   b: {
     c: 'c'
+    d: 'd'
+  }
+}
+
+/* extended state */
+const noExtend: ExtendState<State, never> = {
+  a: 'a',
+  b: {
+    c: 'c',
+    d: 'd'
+  }
+}
+// typings:expect-error
+const noExtendError: ExtendState<State, never> = {
+  a: 'a',
+  b: {
+    c: 'c',
+    d: 'd'
+  },
+  e: 'oops'
+}
+
+const yesExtend: ExtendState<State, { yes: 'we can' }> = {
+  a: 'a',
+  b: {
+    c: 'c',
+    d: 'd'
+  },
+  yes: 'we can'
+}
+// typings:expect-error
+const yesExtendError: ExtendState<State, { yes: 'we can' }> = {
+  a: 'a',
+  b: {
+    c: 'c',
     d: 'd'
   }
 }

--- a/test/typescript/store.ts
+++ b/test/typescript/store.ts
@@ -57,8 +57,8 @@ const funcWithStore = (store: Store<State, DerivedAction>) => {}
 const store: Store<State> = createStore(reducer)
 
 // ensure that an array-based state works
-const arrayReducer = (state = []) => state
-const storeWithArrayState: Store<[]> = createStore(arrayReducer)
+const arrayReducer = (state: any[] = []) => state || []
+const storeWithArrayState: Store<any[]> = createStore(arrayReducer)
 const storeWithPreloadedState: Store<State> = createStore(reducer, {
   a: 'a',
   b: { c: 'c', d: 'd' }

